crates/turborepo-lib/src/process/child.rs

```diff
@@ -379,23 +411,56 @@ impl Child {
         self.pid
     }
 
-    fn stdin(&mut self) -> Option<tokio::process::ChildStdin> {
+    fn stdin(&mut self) -> Option<ChildInput> {
         self.stdin.lock().unwrap().take()
     }
 
-    fn stdout(&mut self) -> Option<tokio::process::ChildStdout> {
+    fn stdout(&mut self) -> Option<ChildOutput<tokio::process::ChildStdout>> {
         self.stdout.lock().unwrap().take()
     }
 
-    fn stderr(&mut self) -> Option<tokio::process::ChildStderr> {
+    fn stderr(&mut self) -> Option<ChildOutput<tokio::process::ChildStderr>> {
         self.stderr.lock().unwrap().take()
     }
 
     /// Wait for the `Child` to exit and pipe any stdout and stderr to the
     /// provided writer.
     pub async fn wait_with_piped_outputs<W: Write>(
         &mut self,
-        mut stdout_pipe: W,
+        stdout_pipe: W,
+    ) -> Result<Option<ChildExit>, std::io::Error> {
+        let stdout_lines = self.stdout();
+        let stderr_lines = self.stderr();
+
+        match (stdout_lines, stderr_lines) {
+            (
+                Some(ChildOutput::Concrete(stdout_lines)),
+                Some(ChildOutput::Concrete(stderr_lines)),
+            ) => {
+                self.wait_with_piped_async_outputs(
+                    stdout_pipe,
+                    Some(BufReader::new(stdout_lines)),
+                    Some(BufReader::new(stderr_lines)),
+                )
+                .await
+            }
+            // If using tokio to spawn tasks we should always be spawning with both stdout and
+            // stderr piped Being in this state indicates programmer error
+            (Some(ChildOutput::Concrete(_)), None) | (None, Some(ChildOutput::Concrete(_))) => {
```

Comment:
[learning rust q]: is this idiomatic Rust? Reads very weird to use a `match` for this. In JS/other scripted languages it would be guard clauses like this.

```js
if (!stdout_lines) throw new Error()
if (!stderr_lines) throw new Error()

// proceed
```

Reply:
Yeah, this wasn't super idiomatic due to how I was representing the possible child IO combinations which lead to a lot of copy/past in the match patterns.

Rust has `let else` expressions which can do the JS style guards e.g.
```
let Some(val) = x else {
    // this branch will only be executed if x doesn't match the pattern
    return Err("expected x to be present");
};
```